File: 55-jump-game/jump-game.py
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        n=len(nums)
        if n==1:
            return True
        max_reach=0
        for i in range(n):
            if i>max_reach:
                return False
            max_reach=max(max_reach,i+nums[i])
            if max_reach >=n-1:
                return True
        return False
        # dp=[None]*n
        # def check(i):
        #     if i==n-1:
        #         return True
        #     if i>=n:
        #         return False
        #     if dp[i] is not None:
        #         return dp[i]
        #     for k in range(1,nums[i]+1):
        #         if check(i+k):
        #             dp[i]=True
        #             return True
        #     dp[i]=False
        #     return False
        # return check(0)
        
        


        

File: 55-jump-game/jump-game.cpp
class Solution {
public:
    vector<int> dp;
    bool solve(vector<int>& nums,int i){
        int n=nums.size();
        if(i==n-1) return true;
        if(i>=n) return false;
        if(dp[i]!=-1) return dp[i];
        for(int j=1;j<=nums[i];j++){
            if(solve(nums,i+j)) return dp[i]=true;
        }
        return dp[i]=false;
    }
    bool canJump(vector<int>& nums) {
        dp.assign(nums.size(),-1);
        return solve(nums,0);
    }
};